### PR TITLE
[MLOB-5297] Prompt Optimization

### PR DESCRIPTION
**Pull request by tillwf**
Closes [MLOB-5297](https://datadoghq.atlassian.net/browse/MLOB-5297)

[MLOB-5297]: https://datadoghq.atlassian.net/browse/MLOB-5297?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+    tag: "Documentation"
+    text: Create and manage datasets
+  - link: "https://www.datadoghq.com/blog/llm-experiments/"
+    tag: "Blog"
+    text: "Create and monitor LLM experiments with Datadog"
+---
+
+## Overview
+
+Prompt Optimization automates the process of improving LLM prompts through systematic evaluation and AI-powered refinement. Instead of manually testing and tweaking prompts, the Prompt Optimizer uses advanced reasoning models to analyze performance data, identify failure patterns, and suggest targeted improvements.
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
Prompt Optimization automates the process of improving LLM prompts through systematic evaluation and AI-powered refinement. Instead of manually testing and tweaking prompts, you can use Prompt Optimization to analyze performance data, identify failure patterns, and suggest targeted improvements.
```

Using two different capitalized names (_Prompt Optimization_ and _Prompt Optimizer_) suggests to the customer that these are two separate products/concepts. It seems like these are just two names for the same entity, so I recommend strictly using only one.

My suggestion to the second sentence is a grammar fix.

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+
+The system runs your LLM application on a dataset with the current prompt, measures performance using your custom metrics, and then uses a reasoning model (such as OpenAI's o3-mini or Claude 3.5 Sonnet) to analyze the results and generate an improved prompt. This cycle repeats until your target metrics are achieved or the maximum number of iterations is reached.
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]: - What does "the system" refer to? 
- For `such as OpenAI's o3-mini or Claude 3.5 Sonnet` — do we expect to be able to update this line whenever new models become available?

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+
+**Key capabilities:**
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
### Prompt Optimization capabilities
```

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+- Automated prompt improvement through AI-powered analysis
+- Customizable evaluation metrics for domain-specific tasks
+- Built-in stopping conditions to prevent over-optimization
+- Parallel experiment execution for rapid iteration
+- Full integration with LLM Observability for tracking and debugging
+
+**Current support:** Prompt Optimization has been validated on Boolean detection tasks and classification use cases, though the architecture supports any output type including structured data extraction, free-form text generation, and numerical predictions.
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
Prompt Optimization supports Boolean detection tasks and classification use cases. Prompt Optimization's architecture supports any output type, including structured data extraction, free-form text generation, and numerical predictions.
```

The documentation should not discuss things in terms of being "current"—_everything_ in the documentation is implied to be current. For example, if support for Feature X is "coming soon," then documentation would state that "Feature X is not supported".

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+
+## How it works
+
+The Prompt Optimizer automates the iterative refinement process through three core components:
+
+1. **Evaluation System**: Runs your LLM application on a dataset and measures performance using custom metrics you define
+2. **Analysis Engine**: Categorizes results into meaningful groups, presents performance data and examples to an AI reasoning model
+3. **Improvement Loop**: AI generates an improved prompt, the system tests it on the full dataset, compares performance, and repeats
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]: Are these components necessary to explain? I think you can build user trust by showing them what's "under the hood", but the concepts here are never referenced again, and I don't think they build a reader's understanding. The bolding of terms and high placement of the page suggest that these concepts are very necessary for a reader to understand, which I don't think is the case. I would recommend removing this segment. The information here could potentially be used in a blog post or other supplemental content.

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+
+Each optimization iteration:
+1. Tests the current prompt on your entire dataset
+2. Analyzes which examples succeed and fail
+3. Shows diverse labeled examples (good/bad/edge cases) to the reasoning model
+4. Generates an improved prompt based on the analysis
+5. Tests the new prompt and compares performance
+6. Keeps the best prompt and continues or stops based on your conditions
+
+The system always optimizes from the current best prompt, ensuring that temporary setbacks don't derail progress.
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]: Again, what does "the system" refer to?

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+        {
+            "input_data": {"conversation": "User: What's the capital of France?\nAI: London"},
+            "expected_output": True  # True = hallucination detected
+        },
+        # Add more records...
+    ]
+)
+```
+
+**Best practice:** Include diverse examples covering typical cases and edge cases to help ensure robust optimization.
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
To help ensure robust optimization, Datadog recommends that you include diverse examples that cover typical cases and edge cases
```

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+    accuracy = (tp + tn) / (tp + tn + fp + fn)
+
+    return {
+        "precision": precision,
+        "recall": recall,
+        "accuracy": accuracy
+    }
+```
+
+**Scoring function** can be one of your metrics or can combine metrics into a single optimization target:
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]: Having a hard time parsing this sentence — I'm not sure what "**Scoring function** can be one of your metrics" means

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+
+```python
+def compute_score(summary_evaluators):
+    """Higher is better. Combine metrics according to business priorities."""
+    metrics = summary_evaluators['precision_recall_evaluator']['value']
+    # Optimize for both precision and accuracy
+    return metrics['precision'] + metrics['accuracy']
+```
+
+**Labelization function** categorizes results for showing diverse examples to the optimizer:
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
**Labelization functions** categorize results for showing diverse examples to the optimizer:
```

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+    """Categorize results into meaningful groups."""
+    eval_value = individual_result["evaluations"]["accuracy_evaluator"]["value"]
+
+    if eval_value in ("true_positive", "true_negative"):
+        return "CORRECT PREDICTION"
+    else:
+        return "INCORRECT PREDICTION"
+```
+
+The labelization function plays a crucial role in optimization: for each unique label, the optimizer receives one randomly selected example from that category (here `CORRECT PREDICTION` and `INCORRECT PREDICTION`). This means the number of labels directly determines the diversity of examples shown to the reasoning model. **Importantly, the label names themselves should be meaningful and descriptive**, as they are shown directly to the reasoning model—use clear, human-readable labels like "HIGH CONFIDENCE ERROR" or "EDGE CASE FAILURE" rather than codes like "TYPE_A" or "CAT_3". Design your labels to represent the key patterns or hints you want the optimizer to learn from, and keep the cardinality low (fewer than 10 distinct labels) to help ensure focused, actionable feedback.
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
The labelization function plays an important role in optimization: for each unique label, the optimizer receives one randomly selected example from that category (in the example above, the labels are `CORRECT PREDICTION` and `INCORRECT PREDICTION`). This means the number of labels directly determines the diversity of examples shown to the reasoning model. 

**Label names should be meaningful and descriptive**, as they are shown directly to the reasoning model. Use clear, human-readable labels like `HIGH CONFIDENCE ERROR` or `EDGE CASE FAILURE` rather than codes like `TYPE_A` or `CAT_3`. Design your labels to represent the key patterns or hints you want the optimizer to learn from, and keep the cardinality low (fewer than 10 distinct labels) to help ensure focused, actionable feedback.
```

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+print(f"Best prompt: {result.best_prompt}")
+print(f"Best score: {result.best_score}")
+print(f"Achieved in {result.total_iterations} iterations")
+print(f"View in Datadog: {result.best_experiment_url}")
+
+# Get score progression
+print(result.summary())
+```
+
+## Configuration options
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]: How/where are these options set? Is it through the configuration dictionary mentioned in line 325?

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+### Evaluation metrics
+
+- Choose metrics aligned with business objectives (precision vs. recall trade-offs)
+- Use multiple complementary metrics rather than a single aggregate score
+- Weight metrics by business impact in your `compute_score` function
+- Test evaluators independently before running optimization
+
+### Labelization
+
+- Create 2-5 distinct, descriptive labels (e.g., "CORRECT HIGH CONFIDENCE", "INCORRECT EDGE CASE")
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
- Create 2-5 distinct, descriptive labels (for example: `CORRECT HIGH CONFIDENCE`, `INCORRECT EDGE CASE`)
```

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+- Helps ensure balanced label distribution (avoid 95% in one category)
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
- Ensure balanced label distribution (avoid 95% in one category)
```

Is the 95% just an example of what to avoid? Or is it a specific threshold that users should stay below?

(Also, the Vale linter loves asking people to put "helps" in front of "ensure" — but this shouldn't be applied if "ensure" is in the imperative tense. Sorry about that, I've corrected it in the suggestion)

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+- Labels help the optimizer understand different types of successes and failures
```
Comment (cswatt) [inline content/en/llm_observability/experiments/prompt_optimization.md]: This doesn't seem like a best practice that the user can put into action, it's just a statement of fact about labels.

Should it be more like:

```suggestion
- Use labels to help the optimizer understand different types of successes and failures
```

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 2 above)
Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md]: - By `the system` I meant the program or this feature or the Prompt Optimization feature.
Not sure which one is the best
- For the reasoning model, maybe we can skip the examples as it will evolve in time indeed. WDYT?

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 4 above)
Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md]: I will change this for:

```
Prompt Optimization supports any use case where the expected output is known and there is a defined way to score the model’s predictions. Prompt Optimization's architecture supports any output type, including structured data extraction, free-form text generation, and numerical predictions.
```

WDYT?

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 8 above)
Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md]: I wanted to say that like the first example the scoring function can simply be one of the values returned by one of the summary evaluators (here the `precision`) or it can be a combination of multiple metric.
How would you formulate this?

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 11 above)
Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md]: It's true that it is not clear. `max_iterations`, `stopping_condition` are `_prompt_optimization` parameters. `jobs` is the parameter of the function `run` and the other configuration are in the `config` parameter.

Should I create sub sections to be clearer?

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 13 above)
Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md]: The 95% is an example. Should I remove it?

Got it for "ensure" :D

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 2 above)
Comment (cswatt) [reply content/en/llm_observability/experiments/prompt_optimization.md]: - Let's go with `Prompt Optimization runs your LLM Application...`
- I think we should skip the examples

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 4 above)
Comment (cswatt) [reply content/en/llm_observability/experiments/prompt_optimization.md]: Yes that sounds great!

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 13 above)
Comment (cswatt) [reply content/en/llm_observability/experiments/prompt_optimization.md]: Let's phrase it as
```suggestion
- Ensure balanced label distribution (for example, avoid 95% in one category)
```

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 8 above)
Comment (cswatt) [reply content/en/llm_observability/experiments/prompt_optimization.md]: Hmmm maybe `The scoring function can return one of the values returned by one of the summary evaluators (for example, precision) or a combination of multiple metrics` ?

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 11 above)
Comment (cswatt) [reply content/en/llm_observability/experiments/prompt_optimization.md]: Ohhh I see, these are being set in the example code for Step 5. Because it's in a new section, I didn't see that the two are related. I have an idea for how to format this more clearly — when you're done making other changes, I'll make my own edits for this section.

Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md]: Thanks a lot! I made the modifications!

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+        {
+            "input_data": {"conversation": "User: What's the capital of France?\nAI: London"},
+            "expected_output": True  # True = hallucination detected
+        },
+        # Add more records...
+    ]
+)
+```
+
+To help ensure robust optimization, Datadog recommends that you include diverse examples that cover typical cases and edge cases
```
Comment (ncybul) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
To help ensure robust optimization, Datadog recommends that you include diverse examples that cover typical cases and edge cases.
```

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+
+### 2. Define your task function
+
+Implement the function that represents your LLM application. This function receives input data and configuration (including the prompt being optimized):
```
Comment (ncybul) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
Implement the function that represents your LLM application. This function receives input data and the configuration (including the prompt being optimized):
```

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+```
+
+**Scoring function** can return one of the values returned by one of the summary evaluators (for example, precision) or a combination of multiple metrics:
+
+```python
+def compute_score(summary_evaluators):
+    """Higher is better. Combine metrics according to business priorities."""
+    metrics = summary_evaluators['precision_recall_evaluator']['value']
+    # Optimize for both precision and accuracy
+    return metrics['precision']
```
Comment (ncybul) [inline content/en/llm_observability/experiments/prompt_optimization.md]:
```suggestion
    # Optimize for precision
    return metrics['precision']
```

**Comment by cswatt** [comment]
> I left some minor suggestions -- I think generally this may be a bit verbose. The best practices section seems like overkill but I think it's fine to leave it in if it would be helpful.

Thanks for these! Applied all suggested changes in latest commit

**Review by cswatt — APPROVED** [review]

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+       accuracy = (tp + tn) / (tp + tn + fp + fn)
+
+       return {
+           "precision": precision,
+           "recall": recall,
+           "accuracy": accuracy
+       }
+   ```
+
+- **Scoring function** can return one of the values returned by one of the summary evaluators (for example, precision) or a combination of multiple metrics:
```
Comment (CFLJacquet) [inline content/en/llm_observability/experiments/prompt_optimization.md:172]: Is this an prompt optim specific concept?

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 19]
```diff
@@ -0,0 +1,404 @@
...
+```
+
+### 3. Define evaluation functions
+
+Create functions that measure how well your task performs. You need four types of functions:
+
+- **Individual evaluators** measure each output:
+
+   ```python
+   def accuracy_evaluator(input_data, output_data, expected_output):
```
Comment (CFLJacquet) [inline content/en/llm_observability/experiments/prompt_optimization.md]: isn't it more "confusion matrix"?  https://en.wikipedia.org/wiki/Confusion_matrix

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 20]
```diff
@@ -0,0 +1,404 @@
...
+       """Higher is better. Combine metrics according to business priorities."""
+       metrics = summary_evaluators['precision_recall_evaluator']['value']
+       # Optimize for both precision and accuracy
+       return metrics['precision'] + metrics['accuracy']
+   ```
+
+- **Labelization functions** categorize results for showing diverse examples to the optimizer:
+
+   ```python
+   def labelization_function(individual_result):
```
Comment (CFLJacquet) [inline content/en/llm_observability/experiments/prompt_optimization.md:195]: Is this also necessary, or could we have a default?

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 21]
```diff
@@ -0,0 +1,404 @@
...
+        ],
+        response_format=DetectionResult
+    )
+
+    return response.choices[0].message.parsed
+```
+
+### 3. Define evaluation functions
+
+Create functions that measure how well your task performs. You need four types of functions:
```
Comment (CFLJacquet) [inline content/en/llm_observability/experiments/prompt_optimization.md]: Suggested explanation:
To optimize your prompt, you need multiple layers of metric computation:
1. At record level to label the results following the confusion matrix labels (TP/FP, TN, FP)
2. [opt] Aggregate these labels into scores
3. Compute the score you want to optimize the prompt for (e.g precision, recall...)
4. Label the missclassification examples to feed to the prompt

Actually is the "3) compute score step" necessary?

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 22]
```diff
@@ -0,0 +1,404 @@
...
+# Visualize improvement
+import matplotlib.pyplot as plt
+plt.plot(result.get_score_history())
+plt.xlabel('Iteration')
+plt.ylabel('Score')
+plt.title('Prompt Optimization Progress')
+plt.show()
+```
+
+## Best practices
```
Comment (CFLJacquet) [inline content/en/llm_observability/experiments/prompt_optimization.md:377]: could we add a cookbook that a user can just run? 

https://github.com/DataDog/llm-observability/tree/main/experiments/notebooks

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 20 above)
Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md:195]: This can be challenging because there may be multiple evaluators, and it’s not always clear which one should be used to distinguish good examples from bad ones. You may also want to further refine the examples by severity—for instance, very poor examples, acceptable examples, and excellent examples.

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 18 above)
Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md:172]: I’m not sure I fully understand the question. I implemented this mechanism because, when I initially optimized for precision, I realized that the accuracy was very poor in the best iteration. I added this function to explicitly define how the best iteration should be determined.

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 19 above)
Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md]: You're right. The name is not accurate.

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 23]
```diff
@@ -0,0 +1,410 @@
...
+```
+
+### 3. Define evaluation functions
+
+To optimize your prompt, you need multiple layers of metric computation:
+
+- At record level, label the results following the confusion matrix labels (TP, FP, TN, FN)
+- Aggregate these labels across all records to compute intermediate metrics (for example, count the total TPs, FPs, TNs, and FNs, then calculate precision, recall, and accuracy)
+- Compute the final score you want to optimize the prompt for by combining or selecting from the aggregated metrics (for example, return precision alone, or combine precision + accuracy)
+- Label the misclassification examples to provide to the prompt optimizer
```
Comment (tillwf) [inline content/en/llm_observability/experiments/prompt_optimization.md:122]: @CFLJacquet @cswatt 
I added this paragraph. WDYT?

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 22 above)
Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md:377]: Ongoing [PR](https://github.com/DataDog/llm-observability/pull/35/changes#r2741947859)

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 23 above)
Comment (cswatt) [reply content/en/llm_observability/experiments/prompt_optimization.md:122]: - If these are steps that need to be followed in order, let's use numbers instead of bullets
- I would write the first one as `At the record level, label the results with confusion matrix conditions: true positive (TP), false positive (FP), true negative (TN), false negative (FN)`
- Because the list of evaluation functions is immediately below this list, you'll need to add a transitional sentence between the two lists that introduces the evaluation functions

Comment (tillwf) [reply content/en/llm_observability/experiments/prompt_optimization.md:122]: The order does not matter.

What do you think of: "Here are implementation examples:" as a transition sentence?

Comment (cswatt) [reply content/en/llm_observability/experiments/prompt_optimization.md:122]: Sounds good, then we can keep them as `-` bullets.

I think I want a little more specificity, something like "The following examples illustrate how to...."

File: content/en/llm_observability/experiments/prompt_optimization.md [hunk 24]
```diff
@@ -0,0 +1,413 @@
...
+- Customizable evaluation metrics for domain-specific tasks
+- Built-in stopping conditions to prevent over-optimization
+- Parallel experiment execution for rapid iteration
+- Full integration with LLM Observability for tracking and debugging
+
+Prompt Optimization supports any use case where the expected output is known and there is a defined way to score the model's predictions. Prompt Optimization's architecture supports any output type, including structured data extraction, free-form text generation, and numerical predictions.
+
+## Prerequisites
+
+- [`ddtrace`][1] version 4.3.0+
```
Comment (CFLJacquet) [inline content/en/llm_observability/experiments/prompt_optimization.md]: 4.4.0

File: content/en/llm_observability/experiments/prompt_optimization.md (hunk 22 above)
Comment (CFLJacquet) [reply content/en/llm_observability/experiments/prompt_optimization.md:377]: can you link the cookbooks here?
